akka-projection-kafka/src/main/scala/akka/projection/kafka/internal/KafkaSourceProviderImpl.scala

```diff
@@ -45,69 +40,83 @@ import org.apache.kafka.common.TopicPartition
 
-  private val subscription = Subscriptions.topics(topics)
-  private lazy val consumerActor =
-    classicSystem.systemActorOf(KafkaConsumerActor.props(settings), nextConsumerActorName())
-  private lazy val metadataClient = MetadataClient.create(consumerActor, KafkaMetadataTimeout)(executionContext)
+  private[kafka] val partitionHandler = new ProjectionPartitionHandler
+  private val subscription = Subscriptions.topics(topics).withPartitionAssignmentHandler(partitionHandler)
+  private val assignedPartitions = new AtomicReference[Set[TopicPartition]](EmptyTps)
 
-  private def stopMetadataClient(): Unit = {
-    metadataClient.close()
-    classicSystem.stop(consumerActor)
-  }
+  protected[internal] def _source(
+      readOffsets: ReadOffsets,
+      numPartitions: Int): Source[ConsumerRecord[K, V], Consumer.Control] =
+    Consumer
+      .plainPartitionedManualOffsetSource(settings, subscription, getOffsetsOnAssign(readOffsets))
+      .flatMapMerge(numPartitions, {
+        case (_, partitionedSource) => partitionedSource
+      })
 
   override def source(readOffsets: ReadOffsets): Future[Source[ConsumerRecord[K, V], _]] = {
     // get the total number of partitions to configure the `breadth` parameter, or we could just use a really large
     // number.  i don't think using a large number would present a problem.
-    val numPartitionsF = Future.sequence(topics.map(metadataClient.getPartitionsFor)).map(_.map(_.length).sum)
-    numPartitionsF.failed.foreach(_ => stopMetadataClient())
+    val numPartitionsF = metadataClient.numPartitions(topics)
+    numPartitionsF.failed.foreach(_ => metadataClient.stop())
     numPartitionsF.map { numPartitions =>
-      Consumer
-        .plainPartitionedManualOffsetSource(settings, subscription, getOffsetsOnAssign(readOffsets))
-        .flatMapMerge(numPartitions, {
-          case (_, partitionedSource) => partitionedSource
-        })
+      _source(readOffsets, numPartitions)
         .watchTermination()(Keep.right)
         .mapMaterializedValue { terminated =>
-          terminated.onComplete(_ => stopMetadataClient())
+          terminated.onComplete(_ => metadataClient.stop())
         }
     }
   }
 
-  override def extractOffset(record: ConsumerRecord[K, V]): MergeableOffset[Long] = {
-    val key = record.topic() + "-" + record.partition()
-    MergeableOffset(Map(key -> record.offset()))
+  override def extractOffset(record: ConsumerRecord[K, V]): GroupOffsets = GroupOffsets(record)
+
+  override def verifyOffset(offsets: GroupOffsets): OffsetVerification = {
+    if (offsets.partitions.forall(assignedPartitions.get().contains))
+      VerificationSuccess
+    else
+      VerificationFailure(
+        "The offset contains Kafka topic partitions that were revoked or lost in a previous rebalance")
   }
 
+  override def isOffsetMergeable: Boolean = true
+
   private def getOffsetsOnAssign(readOffsets: ReadOffsets): Set[TopicPartition] => Future[Map[TopicPartition, Long]] =
```

Comment:
Could we also add a configurable delay here? as an additional safety margin that an envelope in flight on the revoked side was completed before we read the offsets here? Could be a follow up ticket.

Reply:
https://github.com/akka/akka-projection/issues/197